File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* max;
    bool isValidBST(TreeNode* root) {
        if (root == nullptr) {
            return true;
        }

        bool left = isValidBST(root->left);

        if (!left) {
            return false;
        }

        if (max != nullptr && root->val <= max->val) {
            return false;
        }

        max = root;

        return isValidBST(root->right);
        

    }

};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        currVal = None

        def traverse(curr):
            nonlocal currVal
            if not curr:
                return True
            
            if not traverse(curr.left):
                return False

            if not currVal is None and curr.val <= currVal:
                return False

            currVal = curr.val

            if not traverse(curr.right):
                return False

            return True

        return traverse(root)